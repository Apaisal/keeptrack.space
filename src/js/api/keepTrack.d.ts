/* eslint-disable no-unused-vars */
export declare interface SatObject {
  id: number;
  satId?: number;
  position: { x: number; y: number; z: number };
  static?: boolean;
  missile?: boolean;
  type?: string;
  name?: string;
  intlDes?: string;
  TLE1: string;
  TLE2: string;
  SCC_NUM: string;
  active: boolean;
  C?: string;
  LS?: string;
  LV?: string;
  ON: string;
  OT: number;
  TLE1: string;
  TLE2: string;
  R: string;
  URL?: string;
  O?: string;
  U?: string;
  P?: string;
  LM?: string;
  DM?: string;
  Pw?: string;
  Li?: string | number;
  Con?: string;
  M?: string;
  S1?: string;
  S2?: string;
  S3?: string;
  S4?: string;
  S5?: string;
  S6?: string;
  S7?: string;
  inclination: number;
  lon: number;
  perigee: number;
  apogee: number;
  period: number;
  meanMotion: number;
  semimajorAxis: number;
  eccentricity: number;
  raan: number;
  argPe: number;
  inView: number;
  velocity: {
    total: number;
    x: number;
    y: number;
    z: number;
  };
  getTEARR?: any;
  getAltitude?: any;
  getDirection?: any;
  vmag?: number;
  associates?: any;
  maneuver?: any;
  constellation?: any;
  ORPO?: any;
  FMISSED?: any;
  NOTES?: any;
  TTP?: any;
  isInSun?: any;
  inSun?: any;
  desc?: string;
  marker?: boolean;
  isRadarData?: boolean;
  isInGroup?: boolean;
  missileComplex?: number;
}

export interface SensorObject {
  obsmaxel2?: any;
  obsmaxrange2?: any;
  obsminrange2?: any;
  obsminel2?: any;
  obsmaxaz2?: any;
  obsminaz2?: any;
  alt: number;
  beamwidth?: number;
  changeObjectInterval?: number;
  country: string;
  lat: number;
  linkAehf?: boolean;
  linkWgs?: boolean;
  lon: number;
  name: string;
  alt: number;
  observerGd: {
    lat: number;
    lon: number;
    alt: number;
  };
  obsmaxaz: number;
  obsmaxel: number;
  obsmaxrange: number;
  obsminaz: number;
  obsminel: number;
  obsminrange: number;
  shortName: string;
  staticNum: number;
  sun: string;
  type?: string;
  url?: string;
  volume: boolean;
  zoom: string;
}

export interface Lla {
  lat: number;
  lon: number;
  alt: number;
}

export interface Rae {
  rng: number;
  az: number;
  el: number;
}

export interface InView {
  inView: boolean;
}

export interface EarthObject {
  init: any;
  draw: any;
  program: any;
  update: any;
  drawOcclusion: any;
  earthJ: number;
  earthEra: number;
  lightDirection: any;
  pos: [number, number, number];
  loaded: boolean;
  sunvar: any;
  shader: any;
  specularMap: any;
  imgHiRes: any;
  isHiResReady: boolean;
  nightImgHiRes: any;
  nightImg: any;
  bumpMap: any;
  loadHiResNight: any;
  loadHiRes: any;
  isUseHiRes: boolean;
  lightDirection: [number, number, number];
}

export interface keepTrackApiInterface {
  html: (strings: TemplateStringsArray, ...placeholders: any[]) => string;
  register: (params: { method: string; cbName: string; cb: any }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}
