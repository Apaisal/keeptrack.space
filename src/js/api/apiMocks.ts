import { satObject, sensorObject } from './keepTrack';

export const defaultSat: satObject = {
  active: true,
  SCC_NUM: '00005',
  C: 'ISS',
  LS: 'TTMTR',
  LV: 'Proton-K',
  ON: 'ISS (ZARYA)',
  OT: 1,
  TLE1: '1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991',
  TLE2: '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053',
  R: '99.0524',
  URL: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  O: 'National Aeronautics and Space Administration (NASA)/Multinational',
  U: 'Government',
  P: 'Space Science',
  LM: '',
  DM: '',
  Pw: '',
  Li: 30,
  Con: 'Boeing Satellite Systems (prime)/Multinational',
  M: 'Final size of a Boeing 747; first component.',
  S1: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  S2: '',
  S3: 'http://www.boeing.com/defense-space/space/spacestation/overview/',
  S4: '',
  S5: '',
  S6: '',
  S7: '',
  inclination: 51.6423,
  longitude: 168.5744,
  perigee: 1475,
  apogee: 184.3976,
  period: 313.3642,
  meanMotion: 15.48839820294053,
  semimajorAxis: 21203.40407588,
  eccentricity: 0.00003453,
  raan: 0,
  argPe: 0,
  inview: 1,
  velocity: {
    total: 0,
    x: 0,
    y: 0,
    z: 0,
  },
  getTEARR: () => ({
    lon: 168.5744,
    lat: 51.6423,
    alt: 1475,
  }),
  getAltitude: () => 0,
  getDirection: () => 'N',
};

export const defaultSensor: sensorObject = {
  alt: 0.060966,
  beamwidth: 2,
  changeObjectInterval: 1000,
  country: 'United States',
  lat: 41.754785,
  linkAehf: true,
  linkWgs: true,
  lon: -70.539151,
  name: 'Cape Cod AFS, Massachusetts',
  observerGd: { lat: 0.7287584767123405, lon: -1.2311404365114507, alt: 0.060966 },
  obsmaxaz: 227,
  obsmaxel: 85,
  obsmaxrange: 5556,
  obsminaz: 347,
  obsminel: 3,
  obsminrange: 200,
  shortName: 'COD',
  staticNum: 0,
  sun: 'No Impact',
  type: 'Phased Array Radar',
  url: 'http://www.radartutorial.eu/19.kartei/01.oth/karte004.en.html',
  volume: false,
  zoom: 'leo',
};

export const useMockWorkers = () => {
  class Worker {
    url: any;
    onmessage: (msg: any) => void;
    constructor(stringUrl: any) {
      this.url = stringUrl;
      this.onmessage = () => {
        return {
          data: {}
        }
      };
    }

    postMessage(msg: any) {
      if (msg.dat) {
        this.onmessage({          
          data: {
            extraData: JSON.stringify([defaultSat, defaultSensor]),
            satPos: [0,0,0,0,0,0],
            satVel: [0,0,0,0,0,0],
          }
        });
      }
      this.onmessage({
        data: {
          satPos: [0,0,0,0,0,0],
          satVel: [0,0,0,0,0,0],
        }
      });
    }
  }

  (<any>window).Worker = Worker;
}

export const keepTrackApiStubs = {
  programs: {
    adviceManager: {
      adviceList: {
        satViewDisabled: jest.fn(),
        editSatDisabled: jest.fn(),
        survFenceDisabled: jest.fn(),
        bubbleDisabled: jest.fn(),
        satFOVDisabled: jest.fn(),
        planetariumDisabled: jest.fn(),
        mapDisabled: jest.fn(),
      },
      adviceCount: {
        socrates: 0,
      },
      showAdvice: jest.fn(),
      adviceArray: (<any[]>[]),
    },
    mainCamera: {
      cameraType: {
        current: 1,
        satellite: 1,
        fixedToSat: 2,
      },
      camSnap: jest.fn(),
      changeZoom: jest.fn(),
      latToPitch: jest.fn(),
      longToYaw: jest.fn(),
      autoRotate: jest.fn(),
    },
    ColorScheme: {
      reloadColors: jest.fn(),      
    },
    drawManager: {
      glInit: jest.fn(),
      gl: global.mocks.glMock,
      selectSatManager: {
        selectSat: jest.fn(),
      },
    },    
    dotsManager: {
      starIndex1: 0,
      pickReadPixelBuffer: [
        0,
        0,
        0,
      ],
    },    
    groups: null,
    groupsManager: {
      Canada: null,
      Japan: null,
      France: null,
      China: null,
      India: null,
      Israel: null,
      Russia: null,
      UnitedKingdom: null,
      UnitedStates: null,
      SpaceStations: {
        sats: [
          {satId: 1, SCC_NUM: '25544'},
          {satId: 1, SCC_NUM: '25544'},
        ],
        updateOrbits: jest.fn(),
      },
      clearSelect: jest.fn(),
      createGroup: () => ({
        updateOrbits: jest.fn(),
      }),
      selectGroup: jest.fn(),
    },    
    lineManager: {
      create: jest.fn(),
      removeStars: jest.fn(),
      getLineListLen: jest.fn(),
    },
    missileManager: {
      getMissileTEARR: jest.fn(),
    },
    objectManager: {
      isLaunchSiteManagerLoaded: true,
      launchSiteManager: {
        launchSiteList: [
          'AFWTR'
        ],
      },
      selectedSat: -1,
      setSelectedSat: jest.fn(),
      staticSet: {
        id: 0,
      },
      analSatSet: {
        id: 0,
      },
      radarDataSet: {
        id: 0,
      },
      missileSet: {
        id: 0,
      },
      fieldOfViewSet: {
        id: 0,
      },
    },
    orbitManager: {
      updateOrbitBuffer: jest.fn(),
      clearInViewOrbit: jest.fn(),
      removeInViewOrbit: jest.fn(),
      historyOfSatellitesPlay: jest.fn(),
    },
    satellite: {
      currentTEARR: {
        az: 0,
        el: 0,
        rng: 0,
      },
      sgp4: () => ([0,0,0]),
      setobs: jest.fn(),
      setTEARR: jest.fn(),
      altitudeCheck: jest.fn(),
      degreesLong: (num: number) => num,
      degreesLat: (num: number) => num,
      currentEpoch: () => ['21', '150'],
      getOrbitByLatLon: () => ['1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991', '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053'],
      nextpassList: () => [new Date().getTime(), new Date().getTime() + 1],
      twoline2satrec: () => ({ jdsatepoch: 2458000.5 }),
      getRae: () => [0, 0, 0],
      checkIsInFOV: () => true,
      findCloseObjects: () => '',
      findBestPasses: () => jest.fn(),
      nextpass: () => '2021-09-21 20:00:00Z',
      getDOPsTable: jest.fn(),
    },
    satSet: {
      satCruncher: {
        postMessage: jest.fn(),
      },
      satData: [
        defaultSat,
        defaultSat,
      ],
      getIdFromObjNum: jest.fn(),
      getSensorFromSensorName: () => 0,
      getSatExtraOnly: () => defaultSat,
      getSatFromObjNum : () => defaultSat,
      getSat: () => defaultSat,
      initGsData: jest.fn(),
      setColorScheme: jest.fn(),
      getSatInViewOnly: () => defaultSat,
      default: '',
      insertNewAnalystSatellite: jest.fn(),
    },
    searchBox: {
      hideResults: jest.fn(),
      fillResultBox: jest.fn(),
    },
    sensorFov: {
      enableFovView: jest.fn(),
    },
    sensorManager: {
      checkSensorSelected: () => false,
      currentSensor: {
        alt: 0,
        lat: 0,
        lon: 0,
        shortName: 'COD',
      },
    },
    starManager: {
      init: jest.fn(),
    },
    timeManager: {
      propOffset: 0,
      propRate: 0,
      selectedDate: 0,
      propTime: () => 0,
      getDayOfYear: () => 0,
    },
    uiManager: {
      hideSideMenus: jest.fn(),
      toast: jest.fn(),
      colorSchemeChangeAlert: jest.fn(),
      bottomIconPress: jest.fn(),
      legendColorsChange: jest.fn(),
      doSearch: jest.fn(),
      loadStr: jest.fn(),
      searchToggle: jest.fn(),
      legendMenuChange: jest.fn(),
      clearRMBSubMenu: jest.fn(),
      updateURL: jest.fn(),
    },
    watchlist: {
      updateWatchlist: jest.fn(),
    },
  },
};

keepTrackApiStubs.programs.groupsManager.Canada = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Japan = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.France = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.China = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.India = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Israel = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Russia = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedKingdom = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedStates = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groups = keepTrackApiStubs.programs.groupsManager;