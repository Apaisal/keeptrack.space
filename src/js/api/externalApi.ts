interface keepTrackApiInterface {
  register: (params: { method: string; cbName: string; cb: () => void }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}

const keepTrackApi: keepTrackApiInterface = {
  register: (params: { method: string; cbName: string; cb: () => void }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      // Add the callback
      keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
    } else {
      throw new Error(`Invalid callback "${params.method}"!`);
    }
    return;
  },
  unregister: (params: { method: string; cbName: string }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
        if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
          keepTrackApi.callbacks[params.method].splice(i, 1);
          return;
        }
      }
      // If we got this far, it means we couldn't find the callback
      throw new Error(`Callback "${params.cbName} not found"!`);
    } else {
      // Couldn't find the method
      throw new Error(`Invalid callback "${params.method}"!`);
    }
  },
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      for (let i = 0; i < keepTrackApi.callbacks.selectSatData.length; i++) {
        keepTrackApi.callbacks.selectSatData[i].cb(sat, satId);
      }
    },
    updateSelectBox: (sat: any) => {
      for (let i = 0; i < keepTrackApi.callbacks.updateSelectBox.length; i++) {
        keepTrackApi.callbacks.updateSelectBox[i].cb(sat);
      }
    },
  },
  programs: {
    timeManager: {},
    settingsManager: {},
    ColorScheme: {},
    drawManager: {},
    missileManager: {},
    objectManager: {},
    orbitManager: {},
    photoManager: {},
    satSet: {},
    satellite: {},
    searchBox: {},
    sensorManager: {},
    starManager: {},
    sMM: {},
    uiManager: {},
    uiInput: {},
  },
};

(<any>window).keepTrackApi = keepTrackApi;
export { keepTrackApi };
