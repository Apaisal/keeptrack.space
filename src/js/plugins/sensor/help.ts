import { keepTrackApi } from '@app/js/api/keepTrackApi';

export const helpTitleTextSensors = `Sensors Menu`;

export const helpBodyTextSensors = keepTrackApi.html`The Sensors menu allows you to add and remove sensors from KeepTrack. Sensors are used to track the location of satellites.`;
