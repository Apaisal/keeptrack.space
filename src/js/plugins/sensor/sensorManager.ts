/**
 * /////////////////////////////////////////////////////////////////////////////
 *
 * sensorManager.ts is the primary interface between the user interface and the
 * ground based sensors.
 *
 * http://keeptrack.space
 *
 * @Copyright (C) 2016-2021 Theodore Kruczek
 *
 * Original source code released by James Yoder at https://github.com/jeyoder/ThingsInSpace/
 * under the MIT License. Please reference http://keeptrack.space/license/thingsinspace.txt
 *
 * KeepTrack is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * KeepTrack is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with
 * KeepTrack. If not, see <http://www.gnu.org/licenses/>.
 *
 * /////////////////////////////////////////////////////////////////////////////
 */

import { keepTrackApi } from '@app/js/api/externalApi';
import { SensorObject } from '@app/js/api/keepTrack';
import { objectManager } from '@app/js/objectManager/objectManager.js';
import { sensorList } from '@app/js/plugins/sensor/sensorList';
import { setColorScheme } from '@app/js/satSet/satSet';
import $ from 'jquery';

// Add new callbacks to the list of callbacks in keepTrackApi
keepTrackApi.callbacks.setSensor = [];
keepTrackApi.methods.setSensor = (sensor, id) => {
  keepTrackApi.callbacks.setSensor.forEach((cb) => cb.cb(sensor, id));
};

keepTrackApi.callbacks.resetSensor = [];
keepTrackApi.methods.resetSensor = () => {
  keepTrackApi.callbacks.resetSensor.forEach((cb) => cb.cb());
};

export const checkSensorSelected = () => {
  if (sensorManager.currentSensor.lat != null) {
    return true;
  } else {
    return false;
  }
};
export const setCurrentSensor = (sensor: SensorObject | null) => {
  // If the sensor is null, reset the current sensor
  if (sensor == null) {
    sensor = {
      observerGd: {
        lat: null,
        lon: 0,
        alt: 0,
      },
      alt: 0,
      country: '',
      lat: 0,
      lon: 0,
      name: '',
      obsmaxaz: 0,
      obsmaxel: 0,
      obsmaxrange: 0,
      obsminaz: 0,
      obsminel: 0,
      obsminrange: 0,
      shortName: '',
      staticNum: 0,
      sun: '',
      volume: false,
      zoom: '',
    };
  }

  sensorManager.currentSensor = sensor;
};
export const sensorListLength = () => {
  var sensorListCount = 0;
  for (var sensor in sensorList) {
    if (Object.prototype.hasOwnProperty.call(sensorList, sensor)) {
      sensorListCount++;
    }
  }
  return sensorListCount;
};
export const setSensor = (selectedSensor: SensorObject | string, staticNum: number) => {
  try {
    localStorage.setItem('currentSensor', JSON.stringify([selectedSensor, staticNum]));
  } catch {
    console.warn('Sensor Manager: Unable to set current sensor - localStorage issue!');
  }

  // Run any callbacks
  const { satSet, satellite } = keepTrackApi.programs;
  keepTrackApi.methods.setSensor(selectedSensor, staticNum);

  if (selectedSensor == null && staticNum == null) {
    sensorManager.sensorTitle = '';
    $('#sensor-selected').text(sensorManager.sensorTitle);
    return;
  }
  if (selectedSensor === 'SSN') {
    var allSSNSensors = [];
    for (const sensor in sensorList) {
      if (sensorList[sensor].country === 'United States' || sensorList[sensor].country === 'United Kingdom' || sensorList[sensor].country === 'Norway') {
        allSSNSensors.push(sensorList[sensor]);
      }
    }
    satSet.satCruncher.postMessage({
      setlatlong: true,
      sensor: allSSNSensors,
      multiSensor: true,
    });
    satellite.setobs(allSSNSensors);
    objectManager.setSelectedSat(-1);
    setColorScheme(settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'All Space Surveillance Network Sensors';
    // setTimeout(setColorScheme, 1500, settingsManager.currentColorScheme, true);
  } else if (selectedSensor === 'CapeCodMulti') {
    let multiSensor = [];
    multiSensor.push({
      lat: 41.754785,
      lon: -70.539151,
      alt: 0.060966,
      obsminaz: 347,
      obsmaxaz: 227,
      obsminel: 6,
      obsmaxel: 85,
      obsminrange: 150,
      obsmaxrange: 725,
      volume: false,
    });
    multiSensor.push({
      lat: 41.754785,
      lon: -70.539151,
      alt: 0.060966,
      obsminaz: 347,
      obsmaxaz: 227,
      obsminel: 3,
      obsmaxel: 85,
      obsminrange: 700,
      obsmaxrange: 2450,
      volume: false,
    });
    multiSensor.push({
      lat: 41.754785,
      lon: -70.539151,
      alt: 0.060966,
      obsminaz: 347,
      obsmaxaz: 227,
      obsminel: 3,
      obsmaxel: 85,
      obsminrange: 2200,
      obsmaxrange: 5556,
      volume: false,
    });
    satSet.satCruncher.postMessage({
      setlatlong: true,
      sensor: multiSensor,
      multiSensor: true,
    });
    satellite.setobs(sensorManager.sensorList.COD);
    objectManager.setSelectedSat(-1);
    setColorScheme(settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'Cape Cod Multi Fence Radar';
  } else if (selectedSensor === 'NATO-MW') {
    var natoMWSensors = [];
    natoMWSensors.push(sensorManager.sensorList.BLE);
    natoMWSensors.push(sensorManager.sensorList.CAV);
    natoMWSensors.push(sensorManager.sensorList.CDN);
    natoMWSensors.push(sensorManager.sensorList.COD);
    natoMWSensors.push(sensorManager.sensorList.CLR);
    natoMWSensors.push(sensorManager.sensorList.FYL);
    natoMWSensors.push(sensorManager.sensorList.THL);
    satSet.satCruncher.postMessage({
      setlatlong: true,
      sensor: natoMWSensors,
      multiSensor: true,
    });
    satellite.setobs(natoMWSensors);
    objectManager.setSelectedSat(-1);
    setColorScheme(settingsManager.currentColorScheme, true);
    // setTimeout(setColorScheme, 1500, settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'All North American Aerospace Defense Command Sensors';
  } else if (selectedSensor === 'RUS-ALL') {
    var rusSensors = [];
    rusSensors.push(sensorManager.sensorList.ARM);
    rusSensors.push(sensorManager.sensorList.BAL);
    rusSensors.push(sensorManager.sensorList.GAN);
    rusSensors.push(sensorManager.sensorList.LEK);
    rusSensors.push(sensorManager.sensorList.MIS);
    rusSensors.push(sensorManager.sensorList.OLE);
    rusSensors.push(sensorManager.sensorList.PEC);
    rusSensors.push(sensorManager.sensorList.PIO);
    satSet.satCruncher.postMessage({
      setlatlong: true,
      sensor: rusSensors,
      multiSensor: true,
    });
    satellite.setobs(sensorManager.sensorList.ARM);
    objectManager.setSelectedSat(-1);
    setColorScheme(settingsManager.currentColorScheme, true);
    // setTimeout(setColorScheme, 1500, settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'All Russian Sensors';
  } else if (selectedSensor === 'LEO-LABS') {
    var leolabsSensors = [];
    leolabsSensors.push(sensorManager.sensorList.MSR);
    leolabsSensors.push(sensorManager.sensorList.PFISR);
    leolabsSensors.push(sensorManager.sensorList.KSR);
    satSet.satCruncher.postMessage({
      setlatlong: true,
      sensor: leolabsSensors,
      multiSensor: true,
    });
    satellite.setobs(sensorManager.sensorList.MSR);
    objectManager.setSelectedSat(-1);
    setColorScheme(settingsManager.currentColorScheme, true);
    // setTimeout(setColorScheme, 1500, settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'All LEO Labs Sensors';
  } else if (selectedSensor === 'MD-ALL') {
    var mdSensors = [];
    mdSensors.push(sensorManager.sensorList.COD);
    mdSensors.push(sensorManager.sensorList.BLE);
    mdSensors.push(sensorManager.sensorList.CLR);
    mdSensors.push(sensorManager.sensorList.FYL);
    mdSensors.push(sensorManager.sensorList.THL);
    mdSensors.push(sensorManager.sensorList.HAR);
    mdSensors.push(sensorManager.sensorList.QTR);
    mdSensors.push(sensorManager.sensorList.KUR);
    mdSensors.push(sensorManager.sensorList.SHA);
    mdSensors.push(sensorManager.sensorList.KCS);
    mdSensors.push(sensorManager.sensorList.SBX);
    satSet.satCruncher.postMessage({
      setlatlong: true,
      sensor: mdSensors,
      multiSensor: true,
    });
    satellite.setobs(sensorManager.sensorList.MSR);
    objectManager.setSelectedSat(-1);
    setColorScheme(settingsManager.currentColorScheme, true);
    // setTimeout(setColorScheme, 1500, settingsManager.currentColorScheme, true);
    sensorManager.sensorTitle = 'All Missile Defense Agency Sensors';
  } else {
    for (const sensor in sensorList) {
      // console.log(sensorList[sensor] == selectedSensor);
      if (sensorList[sensor] == selectedSensor || (sensorList[sensor].staticNum === staticNum && typeof staticNum != 'undefined')) {
        sensorManager.selectedSensor = sensorList[sensor];
        // Do For All Sensors
        sensorManager.whichRadar = sensorManager.selectedSensor.shortName;
        satSet.satCruncher.postMessage({
          setlatlong: true,
          sensor: sensorManager.selectedSensor,
        });
        satellite.setobs(sensorManager.selectedSensor);

        $('#sensor-info-title').html("<a class='iframe' href='" + sensorManager.selectedSensor.url + "'>" + sensorManager.selectedSensor.name + '</a>');
        try {
          $('a.iframe').colorbox({
            iframe: true,
            width: '80%',
            height: '80%',
            fastIframe: false,
            closeButton: false,
          });
        } catch (error) {
          console.warn(error);
        }
        $('#sensor-type').html(sensorManager.selectedSensor.type);
        $('#sensor-country').html(sensorManager.selectedSensor.country);
        objectManager.setSelectedSat(-1);
        if (typeof settingsManager.currentColorScheme === 'undefined') throw new Error('settingsManager.currentColorScheme is undefined');
        setColorScheme(settingsManager.currentColorScheme, true);
        // setTimeout(setColorScheme, 1500, settingsManager.currentColorScheme, true);
        sensorManager.sensorTitle = sensorManager.selectedSensor.name;
      }
    }
  }
  $('#sensor-selected').text(sensorManager.sensorTitle);
};
export const drawFov = (sensor: SensorObject) => {
  switch (sensor.shortName) {
    case 'COD':
    case 'BLE':
    case 'CLR':
    case 'THL':
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), sensor.obsminaz, sensor.obsminaz + 120, sensor.obsminel, sensor.obsmaxrange], 'c');
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), sensor.obsminaz + 120, sensor.obsmaxaz, sensor.obsminel, sensor.obsmaxrange], 'c');
      break;
    case 'FYL':
      // TODO: Find actual face directions
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), 0, 120, sensor.obsminel, sensor.obsmaxrange], 'c');
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), 120, 240, sensor.obsminel, sensor.obsmaxrange], 'c');
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), 240, 0, sensor.obsminel, sensor.obsmaxrange], 'c');
      break;
    case 'CDN':
      // NOTE: This will be a bit more complicated later
      keepTrackApi.programs.lineManager.create('scan2', [keepTrackApi.programs.satSet.getSensorFromSensorName(sensor.name), sensor.obsminaz, sensor.obsmaxaz, sensor.obsminel, sensor.obsmaxrange], 'c');
      break;
    default:
      console.debug('Sensor not found');
      break;
  }
};

for (var i = 0; i < Object.keys(sensorList).length; i++) {
  sensorList[Object.keys(sensorList)[i]].staticNum = i;
  if (sensorList[Object.keys(sensorList)[i]].obsmaxrange < 8000) {
    sensorList[Object.keys(sensorList)[i]].zoom = 'leo';
  } else {
    sensorList[Object.keys(sensorList)[i]].zoom = 'geo';
  }
}

export const sensorManager = {
  sensorList: sensorList,
  setSensor: setSensor,
  checkSensorSelected: checkSensorSelected,
  sensorListLength: sensorListLength,
  drawFov: drawFov,
  sensorTitle: '',
  setCurrentSensor: setCurrentSensor,
  curSensorPositon: [0, 0, 0],
  whichRadar: '',
  selectedSensor: null,
  sensorListUS: [sensorList.COD, sensorList.BLE, sensorList.CAV, sensorList.CLR, sensorList.EGL, sensorList.FYL, sensorList.THL, sensorList.MIL, sensorList.ALT, sensorList.ASC, sensorList.CDN],
  currentSensor: {
    observerGd: {
      lat: null,
      lon: 0,
      alt: 0,
    },
    alt: 0,
    country: '',
    lat: 0,
    lon: 0,
    name: '',
    obsmaxaz: 0,
    obsmaxel: 0,
    obsmaxrange: 0,
    obsminaz: 0,
    obsminel: 0,
    obsminrange: 0,
    shortName: '',
    staticNum: 0,
    sun: '',
    volume: false,
    zoom: '',
  },
  defaultSensor: {
    observerGd: {
      lat: null,
      lon: 0,
      alt: 0,
    },
  },
};
