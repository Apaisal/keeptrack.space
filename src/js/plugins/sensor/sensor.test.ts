import { keepTrackApiStubs } from '@app/js/api/apiMocks';
import { keepTrackApi } from '@app/js/api/externalApi';
import * as sensor from '@app/js/plugins/sensor/sensor';

keepTrackApi.programs = { ...keepTrackApi.programs, ...keepTrackApiStubs.programs };
// @ponicode

// @ponicode
describe('sensor.init', () => {
  test('0', () => {
    let result: any = sensor.init();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensor.hideSideMenus', () => {
  test('0', () => {
    let result: any = sensor.hideSideMenus();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensor.selectSatData', () => {
  test('0', () => {
    let result: any = sensor.selectSatData();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensor.uiManagerInit', () => {
  test('0', () => {
    let result: any = sensor.uiManagerInit();
    expect(result).toMatchSnapshot();
  });
});

// @ponicode
describe('sensor.bottomMenuClick', () => {
  test('0', () => {
    sensor.bottomMenuClick('menu-lookanglesmultisite');
    let result: any = sensor.bottomMenuClick('menu-lookanglesmultisite');
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    sensor.bottomMenuClick('menu-sensor-list');
    let result: any = sensor.bottomMenuClick('menu-sensor-list');
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    sensor.bottomMenuClick('menu-sensor-info');
    let result: any = sensor.bottomMenuClick('menu-sensor-info');
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    sensor.bottomMenuClick('menu-lookangles');
    let result: any = sensor.bottomMenuClick('menu-lookangles');
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    sensor.bottomMenuClick('menu-customSensor');
    let result: any = sensor.bottomMenuClick('menu-customSensor');
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    let result: any = sensor.bottomMenuClick('');
    expect(result).toMatchSnapshot();
  });
});
