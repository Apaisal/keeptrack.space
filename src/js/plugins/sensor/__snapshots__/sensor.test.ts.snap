// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sensor.bottomMenuClick 0 1`] = `undefined`;

exports[`sensor.bottomMenuClick 1 1`] = `undefined`;

exports[`sensor.bottomMenuClick 2 1`] = `undefined`;

exports[`sensor.bottomMenuClick 3 1`] = `undefined`;

exports[`sensor.bottomMenuClick 4 1`] = `undefined`;

exports[`sensor.bottomMenuClick 5 1`] = `undefined`;

exports[`sensor.hideSideMenus 0 1`] = `undefined`;

exports[`sensor.init 0 1`] = `undefined`;

exports[`sensor.selectSatData 0 1`] = `undefined`;

exports[`sensor.uiManagerInit 0 1`] = `undefined`;
